File: packages/telemetry/src/preflight/run-preflight.ts
import type { GatePolicyInput, GateRunContext } from '@ff-promo/contracts';
import {
	GateRunContextSchema,
	type PreflightCheck,
	type PreflightReport,
} from '@ff-promo/contracts';
import type { PrometheusClient } from '../client/prometheus-client.js';
import { TelemetryApiError } from '../errors/telemetry-adapter-error.js';
import { buildSampleCountQuery } from '../query/build-promql.js';
import { parseInstantQueryResult } from '../query/parse-response.js';

function check(
	id: PreflightCheck['id'],
	status: 'pass' | 'fail',
	extra?: Partial<PreflightCheck>,
): PreflightCheck {
	return { id, status, ...extra };
}

function requiredSampleSize(policies: GatePolicyInput[]): number {
	return policies.reduce(
		(max, policy) => Math.max(max, policy.minSampleSize ?? 0),
		0,
	);
}

function summarizeFailures(checks: PreflightCheck[]): string {
	const failed = checks.filter((item) => item.status === 'fail');
	return failed.map((item) => item.detail ?? item.id).join('; ');
}

function probePolicy(
	policies: GatePolicyInput[],
	runContext: GateRunContext,
): GatePolicyInput {
	return (
		policies[0] ?? {
			metricType: 'error_rate',
			threshold: 0,
			serviceName: runContext.flagKey,
		}
	);
}

export async function runPreflightChecks(
	client: PrometheusClient,
	policies: GatePolicyInput[],
	runContextInput: GateRunContext,
): Promise<PreflightReport> {
	const runContext = GateRunContextSchema.parse(runContextInput);
	const required = requiredSampleSize(policies);
	const policy = probePolicy(policies, runContext);
	const checks: PreflightCheck[] = [];

	try {
		const [treatmentSampleData, controlSampleData] = await Promise.all([
			client.queryInstant(
				buildSampleCountQuery(policy, runContext, 'treatment'),
			),
			client.queryInstant(
				buildSampleCountQuery(policy, runContext, 'control'),
			),
		]);

		const treatmentSample = parseInstantQueryResult(treatmentSampleData);
		const controlSample = parseInstantQueryResult(controlSampleData);

		checks.push(
			check(
				'metric_flow_treatment',
				treatmentSample.ok ? 'pass' : 'fail',
				treatmentSample.ok
					? undefined
					: { detail: 'No treatment user-scoped metrics found' },
			),
		);
		checks.push(
			check(
				'metric_flow_control',
				controlSample.ok ? 'pass' : 'fail',
				controlSample.ok
					? undefined
					: { detail: 'No control user-scoped metrics found' },
			),
		);
		checks.push(
			check(
				'min_sample_treatment',
				treatmentSample.ok && treatmentSample.value >= required ? 'pass' : 'fail',
				{
					observed: treatmentSample.ok ? treatmentSample.value : undefined,
					required,
					detail:
						treatmentSample.ok && treatmentSample.value >= required
							? undefined
							: 'Treatment sample count below required minimum',
				},
			),
		);
		checks.push(
			check(
				'min_sample_control',
				controlSample.ok && controlSample.value >= required ? 'pass' : 'fail',
				{
					observed: controlSample.ok ? controlSample.value : undefined,
					required,
					detail:
						controlSample.ok && controlSample.value >= required
							? undefined
							: 'Control sample count below required minimum',
				},
			),
		);
		checks.push(
			check(
				'context_kind_user',
				treatmentSample.ok ? 'pass' : 'fail',
				treatmentSample.ok
					? undefined
					: { detail: 'No user context kind series present' },
			),
		);
	} catch (error) {
		if (error instanceof TelemetryApiError) {
			checks.push(
				check('metric_flow_treatment', 'fail', {
					detail: 'Prometheus query failed',
				}),
				check('metric_flow_control', 'fail', {
					detail: 'Prometheus query failed',
				}),
				check('min_sample_treatment', 'fail', { required }),
				check('min_sample_control', 'fail', { required }),
				check('context_kind_user', 'fail', {
					detail: 'Prometheus query failed',
				}),
			);
		} else {
			throw error;
		}
	}

	const status = checks.every((item) => item.status === 'pass') ? 'pass' : 'fail';
	return {
		status,
		checks,
		blockReason: status === 'fail' ? summarizeFailures(checks) : undefined,
	};
}

File: kotlin/modules/telemetry/src/main/kotlin/com/ffpromo/telemetry/preflight/RunPreflight.kt
package com.ffpromo.telemetry.preflight

import com.ffpromo.contracts.GatePolicyInput
import com.ffpromo.contracts.GateRunContext
import com.ffpromo.contracts.MetricType
import com.ffpromo.contracts.PreflightCheck
import com.ffpromo.contracts.PreflightCheckStatus
import com.ffpromo.contracts.PreflightReport
import com.ffpromo.telemetry.client.PrometheusClient
import com.ffpromo.telemetry.errors.TelemetryApiError
import com.ffpromo.telemetry.query.Cohort
import com.ffpromo.telemetry.query.ParseResult
import com.ffpromo.telemetry.query.buildSampleCountQuery
import com.ffpromo.telemetry.query.parseInstantQueryResult
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope

private fun check(
    id: String,
    status: PreflightCheckStatus,
    detail: String? = null,
    observed: Double? = null,
    required: Double? = null,
): PreflightCheck =
    PreflightCheck(
        id = id,
        status = status,
        detail = detail,
        observed = observed,
        required = required,
    )

private fun requiredSampleSize(policies: List<GatePolicyInput>): Int =
    policies.maxOfOrNull { it.minSampleSize ?: 0 } ?: 0

private fun summarizeFailures(checks: List<PreflightCheck>): String =
    checks
        .filter { it.status == PreflightCheckStatus.fail }
        .joinToString("; ") { it.detail ?: it.id }

private fun probePolicy(
    policies: List<GatePolicyInput>,
    runContext: GateRunContext,
): GatePolicyInput =
    policies.firstOrNull() ?: GatePolicyInput(
        metricType = MetricType.error_rate,
        threshold = 0.0,
        serviceName = runContext.flagKey,
    )

private fun validateRunContext(runContext: GateRunContext): GateRunContext {
    require(runContext.flagKey.isNotBlank()) { "flagKey is required" }
    require(runContext.treatmentVariationId.isNotBlank()) { "treatmentVariationId is required" }
    require(runContext.controlVariationId.isNotBlank()) { "controlVariationId is required" }
    return runContext
}

suspend fun runPreflightChecks(
    client: PrometheusClient,
    policies: List<GatePolicyInput>,
    runContextInput: GateRunContext,
): PreflightReport {
    val runContext = validateRunContext(runContextInput)
    val required = requiredSampleSize(policies)
    val policy = probePolicy(policies, runContext)
    val checks = mutableListOf<PreflightCheck>()

    try {
        val (treatmentSampleData, controlSampleData) = coroutineScope {
            val treatment = async {
                client.queryInstant(buildSampleCountQuery(policy, runContext, Cohort.treatment))
            }
            val control = async {
                client.queryInstant(buildSampleCountQuery(policy, runContext, Cohort.control))
            }
            treatment.await() to control.await()
        }

        val treatmentSample = parseInstantQueryResult(treatmentSampleData)
        val controlSample = parseInstantQueryResult(controlSampleData)

        checks += check(
            id = "metric_flow_treatment",
            status = if (treatmentSample is ParseResult.Ok) PreflightCheckStatus.pass else PreflightCheckStatus.fail,
            detail = if (treatmentSample is ParseResult.Ok) null else "No treatment user-scoped metrics found",
        )
        checks += check(
            id = "metric_flow_control",
            status = if (controlSample is ParseResult.Ok) PreflightCheckStatus.pass else PreflightCheckStatus.fail,
            detail = if (controlSample is ParseResult.Ok) null else "No control user-scoped metrics found",
        )
        checks += check(
            id = "min_sample_treatment",
            status = if (treatmentSample is ParseResult.Ok && treatmentSample.value >= required) {
                PreflightCheckStatus.pass
            } else {
                PreflightCheckStatus.fail
            },
            observed = (treatmentSample as? ParseResult.Ok)?.value,
            required = required.toDouble(),
            detail = if (treatmentSample is ParseResult.Ok && treatmentSample.value >= required) {
                null
            } else {
                "Treatment sample count below required minimum"
            },
        )
        checks += check(
            id = "min_sample_control",
            status = if (controlSample is ParseResult.Ok && controlSample.value >= required) {
                PreflightCheckStatus.pass
            } else {
                PreflightCheckStatus.fail
            },
            observed = (controlSample as? ParseResult.Ok)?.value,
            required = required.toDouble(),
            detail = if (controlSample is ParseResult.Ok && controlSample.value >= required) {
                null
            } else {
                "Control sample count below required minimum"
            },
        )
        checks += check(
            id = "context_kind_user",
            status = if (treatmentSample is ParseResult.Ok) PreflightCheckStatus.pass else PreflightCheckStatus.fail,
            detail = if (treatmentSample is ParseResult.Ok) null else "No user context kind series present",
        )
    } catch (error: TelemetryApiError) {
        checks += listOf(
            check("metric_flow_treatment", PreflightCheckStatus.fail, detail = "Prometheus query failed"),
            check("metric_flow_control", PreflightCheckStatus.fail, detail = "Prometheus query failed"),
            check("min_sample_treatment", PreflightCheckStatus.fail, required = required.toDouble()),
            check("min_sample_control", PreflightCheckStatus.fail, required = required.toDouble()),
            check("context_kind_user", PreflightCheckStatus.fail, detail = "Prometheus query failed"),
        )
    }

    val status = if (checks.all { it.status == PreflightCheckStatus.pass }) {
        PreflightCheckStatus.pass
    } else {
        PreflightCheckStatus.fail
    }

    return PreflightReport(
        status = status,
        checks = checks,
        blockReason = if (status == PreflightCheckStatus.fail) summarizeFailures(checks) else null,
    )
}
